src/main/java/it/rate/webapp/services/RatingService.java

```diff
@@ -88,4 +93,52 @@ private Criterion getCriterion(Long criterionId) {
         .findById(criterionId)
         .orElseThrow(InvalidCriterionDetailsException::new);
   }
+
+  public List<UserRatedInterestDTO> getAllUserRatedInterestDTOS(AppUser appUser) {
+    List<Rating> userRatings = ratingRepository.findAllByAppUser(appUser);
+
+    Map<Interest, Map<Place, List<Rating>>> ratingsByInterestAndPlace =
+        userRatings.stream()
+            .collect(
+                Collectors.groupingBy(
+                    rating -> rating.getPlace().getInterest(),
+                    Collectors.groupingBy(Rating::getPlace, Collectors.toList())));
+
+    return ratingsByInterestAndPlace.entrySet().stream()
+        .map(
+            entry -> {
+              Interest interest = entry.getKey();
+              List<UserRatedPlaceDTO> userRatedPlaceDTOs =
+                  entry.getValue().entrySet().stream()
+                      .map(
+                          placeEntry -> {
+                            Place place = placeEntry.getKey();
+                            List<UserRatingDTO> userRatingDTOs =
+                                placeEntry.getValue().stream()
+                                    .map(
+                                        userRating ->
+                                            new UserRatingDTO(
+                                                userRating.getRating(),
+                                                userRating.getCriterion().getName()))
+                                    .collect(Collectors.toList());
+
+                            return new UserRatedPlaceDTO(
+                                place.getId(),
+                                place.getName(),
+                                place.getAverageRating() / 2,
+                                userRatingDTOs);
+                          })
+                      .sorted(Comparator.comparingDouble(UserRatedPlaceDTO::avgRating).reversed())
+                      .collect(Collectors.toList());
+
+              return new UserRatedInterestDTO(
+                  interest.getId(),
+                  interest.getName(),
+                  interest.getImageName(),
+                  interest.countLikes(),
+                  userRatedPlaceDTOs);
+            })
+        .sorted(Comparator.comparingInt(UserRatedInterestDTO::likes).reversed())
+        .toList();
+  }
```

Comment:
This should probably be distributed to multiple smaller methods and possibly even services but we can do it later.